Sources/StreamChat/Database/DatabaseContainer.swift

```diff
@@ -222,11 +222,6 @@ class DatabaseContainer: NSPersistentContainer, @unchecked Sendable {
                 try actions(self.writableContext)
                 FetchCache.clear()
 
-                // Refresh the state by merging persistent state and local state for avoiding optimistic locking failure
-                for object in self.writableContext.updatedObjects {
-                    self.writableContext.refresh(object, mergeChanges: true)
-                }
-
```

Comment:
Can you explain why was this needed before, and now it is not? 🤔 What are the chances of this producing some regressions? 🤔

Reply:
Isn't there a chance that Data will be missing or the FRC won't trigger updates?

Reply:
Yes, we need more details on why was this removed.

Reply:
I looked into the history of this and it goes like this:

1. In [2021](https://github.com/GetStream/stream-chat-swift/commit/f22b62e3c73b2b88a652dcb0b404dcb5fa454730) it started as an optimisation change for reducing context changes
```swift
// If you touch ManagedObject and update one of it properties to same value
// Object will be marked as `updated` even it hasn't changed.
// By reseting such objects we remove updates that are not updates.
for object in self.writableContext.updatedObjects {
    if object.changedValues().isEmpty {
        self.writableContext.refresh(object, mergeChanges: false)
    }
}
```
refresh(_:mergeChanges:false) = "Updates the persistent properties of a managed object to use the latest values from the persistent store. If flag is false, the context discards pending changes and the managed object becomes a fault. Upon next access, the context reloads the object’s values from the persistent store or last cached state."

2. In early [2023](https://github.com/GetStream/stream-chat-swift/commit/14773b53decfa46f5682856811a63903786c754d) it gets changed to 
`self.writableContext.refresh(object, mergeChanges: true) `
"If flag is true, the context reloads the object’s property values from the store or the cache. Then the context applies local changes over the newly loaded values. Merging the local values into object always succeeds, and never results in a merge conflict."
I think this was done because change 1 was making FRC to lose track of changes. 
Reasoning:
We use DTOs' willSave to force FRC updates by accessing the "parent" DTO and assigning a property and setting back the same value to the DTO. For example, UserDTO is getting saved and if it happens to be CurrentUserDTO.user, then UserDTO.willSave forces the CurrentUserDTO to change as well. Then FRC observing the CurrentUserDTO picks up the change. Discarding changes in these cases breaks FRC observing which I think the change number 2 tackled. The change in number 2 made sure these updates go through.
From UserDTO:
```swift
if let currentUser = currentUser, !currentUser.hasChanges {
    let fakeNewUnread = currentUser.unreadChannelsCount
    currentUser.unreadChannelsCount = fakeNewUnread
}
```

3. I removed the if check when testing if it fixes these some crashes in CoreData. It didn't.
 
Since the change number 2 some things have happened like we do not use viewContext anymore (except  reading the state in DataStore, DB observers do not use it anymore). All the DB mutations go through the `writableContext`. Therefore, there should never be a case where `writableContext` itself would require to refresh the DTO from the persistent store (because it is the only one changing it). That is the main reason I believe we can remove it.
Interesting side of it is that refreshing an object seems to discard cached data in Core Data (cached data in other contexts) which leads to performance degradation.

Reply:
Didn't we also introduce a state layer context? Will this have an impact on it, have you checked that?

Reply:
For state layer, with similar test, the change is 1.60 s to 0.77 s